src/rgw/rgw_manifest.h

```diff
@@ -0,0 +1,565 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include "rgw_common.h"
+
+class RGWRados;
+class RGWSI_Zone;
+class RGWZoneGroup;
+class RGWZoneParams;
+
+struct rgw_bucket_placement {
+  string placement_rule;
+  rgw_bucket bucket;
+
+  void dump(Formatter *f) const;
+};
+
+class rgw_obj_select {
+  string placement_rule;
+  rgw_obj obj;
+  rgw_raw_obj raw_obj;
+  bool is_raw;
+
+public:
+  rgw_obj_select() : is_raw(false) {}
+  explicit rgw_obj_select(const rgw_obj& _obj) : obj(_obj), is_raw(false) {}
+  explicit rgw_obj_select(const rgw_raw_obj& _raw_obj) : raw_obj(_raw_obj), is_raw(true) {}
+  rgw_obj_select(const rgw_obj_select& rhs) {
+    placement_rule = rhs.placement_rule;
+    is_raw = rhs.is_raw;
+    if (is_raw) {
+      raw_obj = rhs.raw_obj;
+    } else {
+      obj = rhs.obj;
+    }
+  }
+
+  rgw_raw_obj get_raw_obj(const RGWZoneGroup& zonegroup, const RGWZoneParams& zone) const;
+  rgw_raw_obj get_raw_obj(RGWSI_Zone* zone_svc) const;
```

Comment:
i wanted to make this `const RGWSI_Zone*`, but doing that got messy and i split those changes into https://github.com/ceph/ceph/pull/25412